tsl/src/remote/txn_resolve.c

```diff
@@ -85,18 +99,80 @@ remote_txn_heal_data_node(PG_FUNCTION_ARGS)
 		tpc_gid = remote_txn_id_in(id_string);
 		resolution = remote_txn_resolution(foreign_server_oid, tpc_gid);
 
+#ifdef TS_DEBUG
+		/*
+		 * Induce an error in the GID so that the remote side errors out when it tries
+		 * to heal it.
+		 *
+		 * We inject the error by checking the value of the below session variable. Not
+		 * a full GUC, just a tool to allow us to randomly inject error for testing
+		 * purposes. Depending on the value we will inject an error in the GID and also
+		 * additionally change the resolution as per the accepted value:
+		 *
+		 * "commit"  : change GID + set resolution as COMMITTED
+		 * "abort"   : change GID + set resolution as ABORTED
+		 * "unknown" : set resolution as UNKNOWN
+		 *
+		 * Any other setting will not have any effect
+		 */
+		{
+			const char *inject_gid_error =
+				GetConfigOption("timescaledb.debug_inject_gid_error", true, false);
+
+			/* increment the user_id field to cause mismatch in GID */
+			if (inject_gid_error)
+			{
+				if (strcmp(inject_gid_error, "abort") == 0)
+				{
+					tpc_gid->id.user_id++;
+					resolution = REMOTE_TXN_RESOLUTION_ABORT;
+				}
+				else if (strcmp(inject_gid_error, "commit") == 0)
+				{
+					tpc_gid->id.user_id++;
+					resolution = REMOTE_TXN_RESOLUTION_COMMIT;
+				}
+				else if (strcmp(inject_gid_error, "unknown") == 0)
+					resolution = REMOTE_TXN_RESOLUTION_UNKNOWN;
+				/* any other value is simply ignored */
+			}
+		}
+#endif
+		/*
+		 * We don't expect these commands to fail, but if they do, continue and move on to
+		 * healing up the next GID in the list. The ones that failed will get retried if
+		 * they are still around on the datanodes the next time over.
+		 */
 		switch (resolution)
 		{
-			case REMOTE_TXN_RESOLUTION_COMMT:
-				remote_connection_cmd_ok(conn, remote_txn_id_commit_prepared_sql(tpc_gid));
-				resolved++;
+			case REMOTE_TXN_RESOLUTION_COMMIT:
+				if (PQresultStatus(
+						remote_connection_exec(conn, remote_txn_id_commit_prepared_sql(tpc_gid))) ==
+					PGRES_COMMAND_OK)
+				{
+					healed_txn_gids = lappend(healed_txn_gids, id_string);
+					resolved++;
+				}
+				else
+					ereport(WARNING,
+							(errmsg("\"COMMIT PREPARED\" failed. Will be retried later"),
+							 errhint("gid: %s", id_string)));
```

Comment:
This error doesn't give enough context for users when they see it in the log. It is not clear what the prepared transaction is for, which data node is involved, or that his anything to do with multi-node. Also, I would recommend to use `errdetail` for general information and hints for for giving information to the user on how to fix the error.

If we say that this will be retried later it might be a bit  misleading since it sounds as it will happen automatically. This is not always true, since a healing policy must be running. I think we can be more clear, something like:

```suggestion
					ereport(WARNING,
							(errmsg("could not commit prepared transaction %s on data node \"%s\"", id_string, remote_connection_node_name(conn)),
							 errhint("To retry, manually run \"COMMIT PREPARED %s\" on the data node or run the healing function again.", id_string));
```

I would also avoid mentioning GID, as this is an internal name and generally the PG documentation just calls this `transaction_id`.

Btw, what is the behavior of running `COMMIT PREPARED` yourself on the transaction ID? I guess that would work too?


Finally, since this is called in a loop, I am wondering if we risk flooding the output/log with WARNINGs in case of a data node having problems. There might be many prepared transactions to heal on a node. 

One idea is to collect the IDs in the loop and then print only one warning outside the loop that lists all IDs (unless there are a huge amount of transactions that failed).

Reply:
I didn't want to print the `id_string` in the main ereport to get a static report in the test output which is used/visible in the test runs. Will modify the errhint as per your suggestion.

Yes, if you run `COMMIT PREPARED` manually then that will work as well.
You cannot have more than `max_prepared_transactions` in the system which is typically not a very large value. Also, typically there are NOT very many pending prepared transactions in case a datanode is behaving sanely. So I don't expect a lot of entries being obtained by the healing function.